### PR TITLE
[build-tool] Create abstraction for build tools support

### PR DESCRIPTION
**Pull request by seanyen**
* Create abstraction for ROS build tools support. (See https://design.ros2.org/articles/build_tool.html for why I chose `build tool` as the terminology.)

### CODE REVIEW

File: src/buildtool/catkin-make.ts [hunk 1]
```diff
@@ -1,43 +1,49 @@
...
+
+import * as vscode from "vscode"
+
+import * as extension from "../extension"
+import * as common from "./common"
+
+/**
+ * Provides catkin_make build and test tasks.
+ */
+export class CatkinMakeProvider implements vscode.TaskProvider {
```
Comment (kejxu) [inline src/buildtool/catkin-make.ts]: nit: ideally, this should be able to cover both `catkin_make` and `catkin_make_isolated`; suggest rename class to `CatkinProvider` and rename file to `catkin.ts`

File: src/buildtool/buildtool.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    protected _registerTaskProvider(): vscode.Disposable {
+        return null;
+    }
+
+    protected async _createPackage(): Promise<void> {
+        return;
+    }
+}
+
+class CatkinCmakeBuildTool extends BuildTool {
```
Comment (kejxu) [inline src/buildtool/buildtool.ts]: nit: similar to the other comment, suggest renaming this to `CatkinBuildTool`

File: src/extension.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ import * as pfs from "./promise-fs";
 import * as telemetry from "./telemetry-helper";
 import * as vscode_utils from "./vscode-utils";
 
-import * as catkin from "./catkin/catkin";
+import * as buildtool from "./buildtool/buildtool"
```
Comment (kejxu) [inline src/extension.ts]: nit: `build-tool` and `build-tool.ts`

Comment (seanyen) [reply src/extension.ts]: Renamed.

File: src/buildtool/catkin-make.ts (hunk 1 above)
Comment (seanyen) [reply src/buildtool/catkin-make.ts]: I would suggest to decide it when the pull request to support `catkin_make_isolated` is out. Despite they shared the similar names, the arguments and functionality are little bit different and it is hard to tell if we are going to use single `CatkinProvider` for them.

**Comment by seanyen** [comment]
I rebased it to the latest and addressed some feedback.

**Pull request merged**